### PR TITLE
OKTA-335724 .NET Auth SDK > Revisit doc build and auto-publish

### PR DESCRIPTION
**Pull request by andriizhegurov-okta**
## Issue \#

https://oktainc.atlassian.net/browse/OKTA-335724

## Code

- [ ] Unit test(s)
- [ ] Implementation

## Desired behavior

API reference docs are generated on push/merge to master branch or when a new version tag is added.

## Additional Context

For the generation to work following variables should be defined in Travis build environment:
$DEPLOY_DOCS - set to true 
$GITHUB_TOKEN - set to a proper value from github

### CODE REVIEW

File: build.cake [hunk 1]
```diff
@@ -101,6 +119,68 @@ Task("Info")
     Information("Building using {0} version of Cake", cakeVersion);
 });
 
+Task("BuildDocs")
+.IsDependentOn("Build")
+.Does(() =>
+{
+    StartProcess(Context.Tools.Resolve("docfx") ?? Context.Tools.Resolve("docfx.exe"),
```
Comment (bryanapellanes-okta) [inline build.cake:126]: Question: What causes docfx to be available in the Tools context?

File: build.cake [hunk 2]
```diff
@@ -101,6 +119,68 @@ Task("Info")
...
+    if (string.IsNullOrEmpty(travisTag))
+    {
+        Console.WriteLine("TRAVIS_TAG not set, won't copy docs to a tagged directory");
+        return;
+    }
+
+    var taggedVersion = travisTag.TrimStart('v');
+    var tagDocsDirectory = string.Format("./docs/temp/{0}", taggedVersion);
+
+    Information("Copying docs to " + tagDocsDirectory);
```
Comment (bryanapellanes-okta) [inline build.cake:152]: Question: Is this a built in Cake method?

**Review by bryanapellanes-okta — APPROVED** [review]
This looks good to me; had a couple questions for my own curiosity and education.

File: build.cake (hunk 1 above)
Comment (andriizhegurov-okta) [reply build.cake:126]: This line does it https://github.com/okta/okta-auth-dotnet/pull/45/files#diff-4d3634747586b964df9f7712bc3ec669R5

File: build.cake (hunk 2 above)
Comment (andriizhegurov-okta) [reply build.cake:152]: I took this entire method from the older code, I think here this function is:  https://cakebuild.net/api/Cake.Common.Diagnostics/LoggingAliases/C298F7E8

here is an example of output (9015)
image

File: build.cake [hunk 3]
```diff
@@ -101,6 +119,68 @@ Task("Info")
...
+.IsDependentOn("Build")
+.Does(() =>
+{
+    StartProcess(Context.Tools.Resolve("docfx") ?? Context.Tools.Resolve("docfx.exe"), 
+                 "./docs/docfx.json");
+});
+
+Task("CopyDocsToVersionedDirectories")
+.IsDependentOn("BuildDocs")
+.IsDependentOn("CloneExistingDocs")
```
Comment (laura-rodriguez) [inline build.cake:132]: Nit: Let's move this task below `CloneExistingDocs` since this depends on it.

**Review by laura-rodriguez — APPROVED** [review]
I left some questions and one comment, but it looks good overall!

File: build.cake [hunk 4]
```diff
@@ -101,6 +119,68 @@ Task("Info")
...
 
+Task("BuildDocs")
+.IsDependentOn("Build")
+.Does(() =>
+{
+    StartProcess(Context.Tools.Resolve("docfx") ?? Context.Tools.Resolve("docfx.exe"), 
+                 "./docs/docfx.json");
+});
+
+Task("CopyDocsToVersionedDirectories")
```
[the remark this answers is not in the document]
Comment (andriizhegurov-okta) [reply build.cake:130]: Yes, it copies generated docs to the latest folder and (if it's a tagged build) to the version folder.
For now those are accessible via direct links only e.g. https://developer.okta.com/okta-sdk-dotnet/latest/ and  https://developer.okta.com/okta-sdk-dotnet/1.0.0-alpha6/

**Pull request merged**